File: gpoe26-harness-node/src/store/session-store.ts
import { randomUUID } from 'node:crypto';
import type { SessionKind } from '../contract.js';

export interface StoredTurn {
    index: number;
    role: 'user' | 'assistant';
    content: string;
    at: string;
    citations?: string[];
    toolsUsed?: string[];
    compacted?: boolean;
}

export interface StoredSession {
    id: string;
    studentId: string;
    courseId: string;
    kind: SessionKind;
    startedAt: string;
    lastActivityAt: string;
    compactedAt?: string;
    summary?: string;
    turns: StoredTurn[];
}

/**
 * Le magasin de sessions.
 *
 * Deux implémentations derrière une interface — l'idée vient de QM, qui a exactement ce
 * découpage. Elle paie ici plus qu'ailleurs : la suite de conformité peut interroger ce
 * harness sans base de données, donc dans une intégration continue, alors qu'elle vérifie
 * des règles qui n'ont rien à voir avec la persistance.
 */
export interface SessionStore {
    open(studentId: string, courseId: string, kind: SessionKind): Promise<StoredSession>;
    find(id: string): Promise<StoredSession | undefined>;
    append(id: string, turns: StoredTurn[]): Promise<void>;
    compact(id: string, summary: string, upToIndex: number): Promise<void>;
    close(): Promise<void>;
}

/** En mémoire : pour les tests, et pour un développement qui ne veut pas lancer Postgres. */
export class MemorySessionStore implements SessionStore {

    private readonly sessions = new Map<string, StoredSession>();

    async open(studentId: string, courseId: string, kind: SessionKind): Promise<StoredSession> {
        const now = new Date().toISOString();
        const session: StoredSession = {
            id: randomUUID(),
            studentId, courseId, kind,
            startedAt: now,
            lastActivityAt: now,
            turns: [],
        };

        this.sessions.set(session.id, session);
        return session;
    }

    async find(id: string): Promise<StoredSession | undefined> {
        return this.sessions.get(id);
    }

    async append(id: string, turns: StoredTurn[]): Promise<void> {
        const session = this.sessions.get(id);
        if (!session) return;

        session.turns.push(...turns);
        session.lastActivityAt = new Date().toISOString();
    }

    async compact(id: string, summary: string, upToIndex: number): Promise<void> {
        const session = this.sessions.get(id);
        if (!session) return;

        // On marque plutôt qu'on ne supprime : l'historique reste consultable, seul le
        // contexte envoyé au modèle rétrécit. Un élève qui rouvre une session doit
        // pouvoir relire ce qu'il a demandé, même si l'agent ne le rejoue plus.
        for (const turn of session.turns) {
            if (turn.index <= upToIndex) turn.compacted = true;
        }

        session.summary = summary;
        session.compactedAt = new Date().toISOString();
    }

    async close(): Promise<void> {
        this.sessions.clear();
    }
}

File: GPOE26.Harness.Stub/SessionStore.cs
using System.Collections.Concurrent;
using GPOE26.Harness.Contract;

namespace GPOE26.Harness.Stub;

/// <summary>
/// Les sessions du stub, en mémoire.
///
/// Le stub n'a pas à persister quoi que ce soit : il existe pour exercer la suite de
/// conformité, pas pour servir des élèves. Une base ici n'ajouterait qu'une dépendance à
/// démarrer avant de pouvoir lancer les tests.
/// </summary>
public sealed class SessionStore
{
    private readonly ConcurrentDictionary<Guid, StoredSession> _sessions = new();

    public StoredSession Open(Guid studentId, Guid courseId, SessionKind kind)
    {
        var session = new StoredSession(Guid.NewGuid(), studentId, courseId, kind, DateTime.UtcNow);
        _sessions[session.Id] = session;
        return session;
    }

    public StoredSession? Find(Guid id) => _sessions.TryGetValue(id, out var s) ? s : null;
}

public sealed class StoredSession(Guid id, Guid studentId, Guid courseId, SessionKind kind, DateTime startedAt)
{
    private readonly List<TurnDto> _turns = [];
    private readonly Lock _gate = new();

    public Guid Id { get; } = id;
    public Guid StudentId { get; } = studentId;
    public Guid CourseId { get; } = courseId;
    public SessionKind Kind { get; } = kind;
    public DateTime StartedAt { get; } = startedAt;

    public DateTime LastActivityAt { get; private set; } = startedAt;
    public DateTime? CompactedAt { get; private set; }

    public void Record(string role, string content, IReadOnlyList<string>? citations = null)
    {
        lock (_gate)
        {
            _turns.Add(new TurnDto(_turns.Count, role, content, DateTime.UtcNow, citations));
            LastActivityAt = DateTime.UtcNow;
        }
    }

    public void Compact()
    {
        lock (_gate)
        {
            // Compaction du stub : on ne garde que le dernier échange. Le vrai harness
            // résumera ; ce qui compte ici est que la date soit posée et l'historique réduit.
            if (_turns.Count > 2) _turns.RemoveRange(0, _turns.Count - 2);
            CompactedAt = DateTime.UtcNow;
        }
    }

    public IReadOnlyList<TurnDto> Turns
    {
        get
        {
            // ForgetHistory : une session reprise revient vide, ce que le test de reprise
            // doit refuser.
            if (SabotageSettings.Is(Sabotage.ForgetHistory)) return [];

            lock (_gate) return [.. _turns];
        }
    }

    public SessionDto ToDto() => new(
        Id, StudentId, CourseId, Kind, StartedAt, LastActivityAt,
        Turns.Count(t => t.Role == "user"),
        CompactedAt,
        StubTools.For(Kind));
}
